aggregation.go

```diff
@@ -1 +1,281 @@
+				// * returned. However, we don't want count null value of a certain field, as a result we count
+				// * documents w/ non-null value of the target field by "value_count" keyword
+				aggNameStr = "value_count"
+			}
+		case "avg", "sum", "min", "max":
+			if funcExpr.Distinct {
+				err := fmt.Errorf(`esql: aggregation function %v w/ DISTINCT not supported`, aggNameStr)
+				return nil, nil, nil, nil, err
+			}
+			aggTagStr = aggNameStr + "_" + aggTargetStr
+		default:
+			err := fmt.Errorf(`esql: aggregation function %v not supported`, aggNameStr)
+			return nil, nil, nil, nil, err
+		}
+		if _, exist := aggTagSet[aggTagStr]; exist {
+			continue
+		}
+		aggTagSet[aggTagStr] = len(aggTagSet)
+		aggNameSlice = append(aggNameSlice, aggNameStr)
+		aggTargetSlice = append(aggTargetSlice, aggTargetStr)
+		aggTagSlice = append(aggTagSlice, aggTagStr)
+	}
+
+	return aggNameSlice, aggTargetSlice, aggTagSlice, aggTagSet, nil
+}
+
+func (e *ESql) extractSelectedExpr(expr sqlparser.SelectExprs) ([]*sqlparser.FuncExpr, []string, []string, error) {
+	var aggFuncExprSlice []*sqlparser.FuncExpr
+	var colNameSlice, aggNameSlice []string
+	for _, selectExpr := range expr {
+		// from sqlparser's definition, we need to first convert the selectExpr to AliasedExpr
+		// and then check whether AliasedExpr is a FuncExpr or just ColName
+		switch selectExpr.(type) {
+		case *sqlparser.AliasedExpr:
+			aliasedExpr := selectExpr.(*sqlparser.AliasedExpr)
+			switch aliasedExpr.Expr.(type) {
+			case *sqlparser.FuncExpr:
+				funcExpr := aliasedExpr.Expr.(*sqlparser.FuncExpr)
+				aggFuncExprSlice = append(aggFuncExprSlice, funcExpr)
+				aggNameSlice = append(aggNameSlice, sqlparser.String(funcExpr.Exprs))
+			case *sqlparser.ColName:
+				colName := aliasedExpr.Expr.(*sqlparser.ColName)
+				colNameSlice = append(colNameSlice, strings.Trim(sqlparser.String(colName), "`"))
+			default:
+				err := fmt.Errorf(`esql: %T not supported in select body`, aliasedExpr)
+				return nil, nil, nil, err
+			}
+		default:
+		}
+	}
+	return aggFuncExprSlice, colNameSlice, aggNameSlice, nil
+}
+
+func (e *ESql) convertGroupByExpr(expr sqlparser.GroupBy) (dsl string, colNameSet map[string]int, err error) {
+	var groupByStrSlice []string
+	colNameSet = make(map[string]int)
+	for _, groupByExpr := range expr {
+		switch groupByItem := groupByExpr.(type) {
+		case *sqlparser.ColName:
+			colNameStr := groupByItem.Name.String()
+			if _, exist := colNameSet[colNameStr]; !exist {
+				colNameSet[colNameStr] = 1
+				groupByStr := fmt.Sprintf(`{"group_%v": {"terms": {"field": "%v", "missing_bucket": true}}}`, colNameStr, colNameStr)
+				groupByStrSlice = append(groupByStrSlice, groupByStr)
+			}
+		default:
+			err = fmt.Errorf(`esql: GROUP BY %T not supported`, groupByExpr)
+			return "", nil, err
+		}
+	}
+	dsl = strings.Join(groupByStrSlice, ",")
+	// TODO: magic size number, use "after" for pagination
```

Comment:
make sure we note it somewhere apart from code so that we don't forget it.

Reply:
I added some reminder in README.